internal/session/handle_id.go

```diff
@@ -0,0 +1,42 @@
+package session
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ProtonMail/gluon/imap"
+	"github.com/ProtonMail/gluon/internal"
+	"github.com/ProtonMail/gluon/internal/parser/proto"
+	"github.com/ProtonMail/gluon/internal/response"
+)
+
+func prepareServerResponse(info *internal.VersionInfo) *response.IdResponse {
+	return &response.IdResponse{
+		ID: imap.NewIMAPIDFromVersionInfo(info),
+	}
+}
+
+func (s *Session) handleIDGet(ctx context.Context, versionInfo *internal.VersionInfo, tag string, ch chan response.Response) error {
+	ch <- prepareServerResponse(versionInfo)
+	ch <- response.Ok(tag)
+
+	return nil
+}
+
+func (s *Session) handleIDSet(ctx context.Context, versionInfo *internal.VersionInfo, tag string, cmd *proto.IDSet, ch chan response.Response) error {
+	// Update session information
+	s.imapID = imap.NewIMAPIDFromKeyMap(&cmd.Keys)
+
+	// Not logged in or no mailbox selected
+	if s.state != nil {
+		if err := s.state.SetConnMetadataKeyValue(imap.IMAPIDConnMetadataKey, s.imapID); err != nil {
+			ch <- response.Bad(tag, fmt.Sprintf("Failed to store IMAP ID: %v", err))
```

Comment:
You can just directly return the bad response, no need to send it and then return an error afterwards (see eg [here](https://github.com/ProtonMail/gluon/blob/36e208671b623c7c99fe6765cb9d802cc764553c/internal/session/handle_login.go#L21))

Reply:
Done